### PR TITLE
Feature Completed

### PR DESCRIPTION
**Pull request by Aaron-RN**
my_enumerable versions of built in ruby functions completed

### CODE REVIEW

File: my_Enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        else
+          if pattern
+            if pattern === (item)
+              return true
+            end
+          else
+            if item == nil||item == false
+              return true
+            end
+          end
```
Comment (samgaco) [inline my_Enumerable.rb]: ### 1. If at least one of the elements of the evaluated collection is not false, it should return true, example:

expected behaviour:

```
irb(main):168:0> arr = [false, false]
=> [false, false]
irb(main):170:0> arr.any?
=> false
irb(main):171:0> arr = [false, true]
=> [false, true]
irb(main):173:0> arr.any?
=> true

```

currently:

```
irb(main):168:0> arr = [false, false]
=> [false, false]
irb(main):169:0> arr.my_any?
=> true
irb(main):171:0> arr = [false, true]
=> [false, true]
irb(main):172:0> arr.my_any?
=> true


```

File: my_Enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            return false
+          end
+        else
+          if pattern
+            if pattern === (item)
+              return false
+            end
+          else
+            if item == nil||item == false
+              return false
```
Comment (samgaco) [inline my_Enumerable.rb]: ### 2. Similar problem as the one seen in my_any? , expected behavior compared with the current method implemented in your project:

```
irb(main):174:0> arr = [false, false]
=> [false, false]
irb(main):175:0> arr.none?
=> true
irb(main):176:0> arr.my_none?
=> false

```

File: my_Enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      newArray = []
+      self.my_each do |item|
+          if proc
+            newArray << proc.call(item)
+          else
+            newArray << yield(item)
+          end
+      end
+      return newArray
+    end
```
Comment (samgaco) [inline my_Enumerable.rb]: ### 3. my_map should return an enumerator if no block or arguments are given, example:

```
irb(main):179:0> arr=[1,2,3]
=> [1, 2, 3]
irb(main):180:0> arr.map
=> #<Enumerator: [1, 2, 3]:map>
```

File: my_Enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        if operator!=nil && !block_given?
+          finalValue = finalValue.public_send(operator.to_sym, item)
+        else
+          if block_given?
+            finalValue = yield(finalValue, item)
+          end
+        end
+      end
+      return finalValue
+    end
```
Comment (samgaco) [inline my_Enumerable.rb]: ### 4. problems with inject, if we add a number of the argument, the accumulator should be initiated with that argument

```
irb(main):195:0> arr = [1,2,3]
=> [1, 2, 3]
irb(main):196:0> arr.inject(2) { |n, z| n + z }
=> 8
irb(main):197:0> arr.my_inject(2) { |n, z| n + z }
=> 6
irb(main):198:0> (1..3).inject(2) { |prod, n| prod * n}
=> 12
irb(main):199:0> (1..3).my_inject(2) { |prod, n| prod * n}
=> 6
```

---------------------------------

### 5. a symbol should be able to pass through and be applied as a operation through all the objects to which we apply the method


```
irb(main):200:0> (1..3).inject(2, :+) 
=> 8
irb(main):201:0> (1..3).my_inject(2, :+) 
=> 6
irb(main):202:0> (1..3).inject(2, :*) 
=> 12
irb(main):203:0> (1..3).my_inject(2, :*) 
=> 6
irb(main):204:0> (1..3).inject(5, :*) 
=> 30
```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi  @Aaron-RN  :high_brightness: 

I left 5 points below to be addressed, the five of them are required to get the project approved.

Good spirit :+1: 

---------------------

Moving on :arrow_down: 
* **[For your next code review requests]**  Add a short summary of changes in Pull Request description
* **[For your next code review requests]**  Use meaningful names for your Pull Requests
-------------------------
* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: my_Enumerable.rb (hunk 4 above)
Comment (Aaron-RN) [reply my_Enumerable.rb]: Ohhh... I assumed for inject the initial value was linked to the element selected from the range or array....
Not that the initial value was the actual value the inject starts off as...

**Comment by Aaron-RN** [comment]
> * **[For your next code review requests]**  Add a short summary of changes in Pull Request description
> * **[For your next code review requests]**  Use meaningful names for your Pull Requests

Yeah I did this previously... I created an individual feature branch for each method I created and merged into development branch but I was told that the default branch was to be made the development branch and should just join all features into one branch called feature_branch so I redid the repo

**Review by Ispirett — CHANGES_REQUESTED** [review]
Really great  job with this and implemented changes @Aaron-RN  :+1:  

image



There is just I little bug :bug: 

**My_inject**

words =  //=> an array of words
I tried this  `  words.my_inject { proc { |memo, word| memo.length > word.length ? memo : word }`
and got the following error ` undefined method `length' for 0:Integer`

File: my_Enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+          else
+            newArray << yield(item)
+          end
+      end
+      return newArray
+    end
+
+    def my_inject(initial = nil, operator = nil)
+      if initial.class == Symbol
+        operator = initial
```
Comment (samgaco) [inline my_Enumerable.rb]: Your my_inject method is having some odd behavior when tested without arguments or blocks:
```

irb(main):153:0> [1,2,3,4,5].inject(:+)
=> 15
irb(main):154:0> [1,2,3,4,5].my_inject(:+)
=> 12
irb(main):155:0> [10,50,1,2,3].inject(:+)
=> 66
irb(main):156:0> [10,50,1,2,3].my_inject(:+)
=> 66
irb(main):157:0> [1,2,3].inject(:+)
=> 6
irb(main):158:0> [1,2,3].my_inject(:+)
=> 3
irb(main):159:0> [1,2,4].my_inject(:+)
=> 7
irb(main):160:0> [1,2,4,5].inject(:+)
=> 12
irb(main):161:0> [1,2,4,5].my_inject(:+)
=> 12
irb(main):162:0> [1,2,4].my_inject(:+)
=> 7
irb(main):163:0> [1,2,4].inject(:+)
=> 7
irb(main):164:0> [1,2,3].inject(:+)
=> 6
irb(main):165:0> [1,2,3].my_inject(:+)
=> 3
 
```

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @Aaron-RN :high_brightness: 

I pointed you out in the comments below what it seems to be the last problem with your code, please :pray: address it and submit back for another code review once it's done.

Almost there :100: 

-----------------------
* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

File: my_Enumerable.rb (hunk 5 above)
Comment (Aaron-RN) [reply my_Enumerable.rb]: Good spot I figured that last line I added might cause issues... Got it sorted though.

File: my_Enumerable.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    my_each do |item|
+      if block_given?
+        return false unless yield(item)
+      elsif pattern
+          return false unless pattern === item
+        else
+          return false if item.nil? || item == false
+        end
+      end
+    end
```
Comment (samgaco) [inline my_Enumerable.rb:41]: Be careful with this, you have an extra `end` over here.

it should look like:
```

  def my_all?(pattern = nil)
    my_each do |item|
      if block_given?
        return false unless yield(item)
      elsif pattern
          return false unless pattern === item
      else
          return false if item.nil? || item == false
      end
    end
    true
  end
```

**Review by samgaco — APPROVED** [review]
Hi @Aaron-RN :high_brightness: 

The project is approved, you may move on :+1: 


---------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:

**Pull request merged**